### ISSUE TITLE
Fix test failures and errors on Windows

### ISSUE THREAD

**Issue by slarse** [INRIA/spoon#3752]
With #3746 we're adding a new build for Java 1.8 on Windows. There are 53 test failures and 4 errors that are specific to Windows that should be fixed. All of these start with the following line:

```java
assumeNotWindows(); // FIXME Make test case pass on Windows
```

**Comment by monperrus** [comment]
@slarse the Windows CI build fails on master: https://github.com/INRIA/spoon/actions/runs/484260233

Would you have a look? Thanks!

**Comment by slarse** [comment]
The failing test was on a branch that didn't have the Windows build, and was merged very close to it. Many of the sniper printer tests don't work well on Windows, it probably has something to do with newlines not being correctly printed with LF line endings (just a hunch).

I'm disabling that test for the time being, and then when someone gets around to fixing Windows test failures they can deal with it. It's very likely that most, perhaps all, of the sniper test failures are caused by the same issue.

**Comment by slarse** [comment]
Just an FYI, I don't intend to be the one to fix these test errors. Anyone with a Windows dev environment would be a better candidate.

@MartinWitt Did you feel like doing it?

**Comment by MartinWitt** [comment]
On windows we have 2 problems, paths and carriage returns. The path problem is fixed by using the java api and not comparing strings. For the carriage return problem exist multiple solutions. Loo
```java
	@Test
	public void testTextBlockCreation(){
		// contract: Test creation of TextBlock and prettyprinting
		Factory factory = getSpoonFactory();
		CtClass<?> c = Launcher.parseClass("class Test{public String m1(){String s = \"\";}}");
		CtBlock<?> body = c.getMethod("m1").getBody();
		CtReturn ret = factory.createReturn();
		ret.setValueByRole(CtRole.EXPRESSION, factory.createTextBlock("Hello, \"World\"!\nTesting\n\tTabs"));
		body.insertEnd(ret);
		assertEquals(
				"class Test {\n"
				+ "    public String m1() {\n"
				+ "        String s = \"\";\n"
				+ "        return \"\"\"\n"
				+ "        Hello, \"World\"!\n"
				+ "        Testing\n"
				+ "        	Tabs\"\"\";\n"
				+ "    }\n"
				+ "}",
				c.toString()
		);
	}
```
If we take a look at this testcase, we have 3 solutions for the common problem with carriage returns.
### Remove "\r"
Adding a `.replaceAll("\r", "")` for the `toString` fixes the testcase, but not the core problem
### Write own assertions
We could write a String assertion like `assertSameContent` which internally removes "\r". Would take more effort but keeps testcases clean and simple. Works as long as no one tries to use `"\r"` randomly in his code(Why would someone do that?).
### Fix/Improve Printer
I found no option to set line endings in the printer. Setting the line endings in printer to `\n` would remove almost all test problems and adding a flag for windows line endings would users still give the freedom to chose. Or we keep line endings as `System.lineSeparator`, but add the option to set them. But for printer/sniper changes is @monperrus the domain expert in guessing how hard a change is.

Would is your preferred solution?

**Comment by slarse** [comment]
> The path problem is fixed by using the java api and not comparing strings.

Is this internal in Spoon, or just in the tests?

> Works as long as no one tries to use "\r" randomly in his code(Why would someone do that?).

Off the top of my head: regex.

> Setting the line endings in printer to \n would remove almost all test problems and adding a flag for windows line endings would users still give the freedom to chose. 

There's a method `DefaultJavaPrettyPrinter.setLineSeparator(String)`: https://github.com/INRIA/spoon/blob/bf5b116d176f26fa0f66a14e39c6db237bf70298/src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java#L237-L244

That method isn't exposed beyond the API of the `DefaultJavaPrettyPrinter`, and so for some test cases it might be very inconvenient to set it. Perhaps it needs to be possible to set it in e.g. the environment.

> But for printer/sniper changes is @monperrus the domain expert in guessing how hard a change is.

At least for the sniper, I'm probably the one to consult, as I'm afaik the only one currently working on it.

**Comment by slarse** [comment]
As an example of it being a bit awkward to access the `DefaultJavaPrettyPrinter.setLineSeparator()` method, here's what it could look like for the `testTextBlockCreation()` test case above:

```java
	@Test
	public void testTextBlockCreation(){
		assumeNotWindows(); // FIXME Make test case pass on Windows
		// contract: Test creation of TextBlock and prettyprinting
		Factory factory = getSpoonFactory();
		CtClass<?> c = Launcher.parseClass("class Test{public String m1(){String s = \"\";}}");
		CtBlock<?> body = c.getMethod("m1").getBody();
		CtReturn ret = factory.createReturn();
		ret.setValueByRole(CtRole.EXPRESSION, factory.createTextBlock("Hello, \"World\"!\nTesting\n\tTabs"));
		body.insertEnd(ret);

		c.getFactory().getEnvironment().setPrettyPrinterCreator(() -> {
			DefaultJavaPrettyPrinter printer = (DefaultJavaPrettyPrinter)
					c.getFactory().getEnvironment().createPrettyPrinterAutoImport();
			printer.setLineSeparator("\n");
			return printer;
		});

		assertEquals(
				"class Test {\n"
				+ "    public String m1() {\n"
				+ "        String s = \"\";\n"
				+ "        return \"\"\"\n"
				+ "        Hello, \"World\"!\n"
				+ "        Testing\n"
				+ "        	Tabs\"\"\";\n"
				+ "    }\n"
				+ "}",
				c.toString()
		);
	}
```

We could of course define helpers for these things, but as this is fundamental to using Spoon on Windows it seems like it should be a bit smoother. A user on Windows probably _doesn't_ want CRLF line endings in the code anyway, so it should preferably be easy to set.

**Comment by MartinWitt** [comment]
> > The path problem is fixed by using the java api and not comparing strings.
> 
> Is this internal in Spoon, or just in the tests?

Only for tests, sometimes we compare if a path ends with another path.  And all paths in testcases are obviously written as unix paths but the file path from a written file is a windows path.
> 
> > Works as long as no one tries to use "\r" randomly in his code(Why would someone do that?).
> 
> Off the top of my head: regex.

Good point that means we would create a new problem some1 in the future maybe has to fix. That makes the solution less preferable.
> 
> > Setting the line endings in printer to \n would remove almost all test problems and adding a flag for windows line endings would users still give the freedom to chose.
> 
> There's a method `DefaultJavaPrettyPrinter.setLineSeparator(String)`:
> 
>
>https://github.com/INRIA/spoon/blob/bf5b116d176f26fa0f66a14e39c6db237bf70298/src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java#L237-L244
> 
> That method isn't exposed beyond the API of the `DefaultJavaPrettyPrinter`, and so for some test cases it might be very inconvenient to set it. Perhaps it needs to be possible to set it in e.g. the environment.

Is there any risk to add a flag in environment? Could this option produce any problems? 
To keep it backwards compatible, we could use the default als system.lineSeparator() and create an enum like this
```java
enum LineEndings() {
Unix("\n"),Windows("\r\n");
}
```
, `setLineEndings(LineEnding ending)` in environment and use String getLineEnding() in printer etc.
> 
> > But for printer/sniper changes is @monperrus the domain expert in guessing how hard a change is.
> 
> At least for the sniper, I'm probably the one to consult, as I'm afaik the only one currently working on it.

Oh sry it's hard to follow all changes. Are there any problems in the sniper if we allow changing line endings?

**Comment by slarse** [comment]
> Oh sry it's hard to follow all changes.

Oh don't worry about it, I didn't mean it as "I'm offended" :)

> Are there any problems in the sniper if we allow changing line endings?

Well, the sniper printer will ignore it, it always detects the line separator when calculating new output, here:

https://github.com/INRIA/spoon/blob/bf5b116d176f26fa0f66a14e39c6db237bf70298/src/main/java/spoon/support/sniper/SniperJavaPrettyPrinter.java#L136

So the only problem is perhaps that it would be a bit non-obvious for users of the sniper printer. But I think that's acceptable, if we're diligent we could even issue a warning about an "incompatible setting" or something.

**Comment by monperrus** [comment]
FYI, info from Github

> Windows Server 2022 became [generally available on GitHub-hosted runners](http://app.github.media/e/er?s=88570519&lid=2917&elqTrackId=78b02f6b9944420daba8469c480f92c3&elq=4a5569a662464826aa9a78629cc5aca9&elqaid=2344&elqat=1) in November 2021. Starting this week and over the next 8 weeks, jobs using the `windows-latest` runner label will migrate from Windows Server 2019 to Windows Server 2022. During migration, you can determine if your job has migrated by viewing the `Virtual Environment` information in the `Set up job` step of your logs.
GitHub

> The Windows Server 2022 runner image has different tools and tool versions than Windows Server 2019. See the full list of changed software [here.](http://app.github.media/e/er?s=88570519&lid=2916&elqTrackId=250877cd68354dac8626d1abd34428b6&elq=4a5569a662464826aa9a78629cc5aca9&elqaid=2344&elqat=1)

**Comment by MartinWitt** [comment]
Btw this issue seems fixed? We have only one unused import for `assumeNotWindows` left? https://github.com/INRIA/spoon/search?q=assumeNotWindows

**Comment by monperrus** [comment]
yes, let's remove the unused import and close the issue.
